File: android/app/src/main/kotlin/com/kapybara/kapynotes/SystemRegion.kt
package com.kapybara.kapynotes

import android.content.res.Resources
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodChannel

/**
 * Backs the `kapynotes/region` channel: the country this device is set to.
 *
 * Read from the *system* configuration rather than the activity's. Android 13
 * lets a single app be given a language of its own, and an app told to speak
 * Hindi has not moved to India — the phone's own locale list is the one that
 * says where its owner is.
 *
 * Android is the platform where region and language genuinely are one setting,
 * so this usually agrees with what Flutter already reports. It is here so that
 * the Dart side has one question with one answer everywhere; see
 * lib/core/system_region.dart for the platforms where they come apart.
 */
object SystemRegion {
    private const val CHANNEL = "kapynotes/region"

    fun register(messenger: BinaryMessenger): MethodChannel {
        val channel = MethodChannel(messenger, CHANNEL)
        channel.setMethodCallHandler { call, result ->
            when (call.method) {
                "region" -> result.success(current())
                else -> result.notImplemented()
            }
        }
        return channel
    }

    /** Null where the system names no country, which Dart reads as "use the language". */
    private fun current(): String? {
        val locales = Resources.getSystem().configuration.locales
        if (locales.isEmpty) return null
        return locales.get(0).country.ifEmpty { null }
    }
}

File: macos/Runner/SystemRegion.swift
import FlutterMacOS
import Foundation

/// Backs the `kapynotes/region` channel: the country this device is set to.
///
/// `Locale.current` carries the Region setting even when the app is being read
/// in another language — on a Mac set to English (US) with Region India its
/// identifier is `en_US@rg=inzzzz`, and `region` is `IN`. The preferred
/// languages list that Flutter builds its locale from says `en-US` and knows
/// nothing about the `IN`, which is why the Dart side has to come here.
enum SystemRegion {
  static let channelName = "kapynotes/region"

  static func register(with messenger: FlutterBinaryMessenger) -> FlutterMethodChannel {
    let channel = FlutterMethodChannel(name: channelName, binaryMessenger: messenger)
    channel.setMethodCallHandler { call, result in
      guard call.method == "region" else {
        result(FlutterMethodNotImplemented)
        return
      }
      result(current())
    }
    return channel
  }

  /// Nil rather than a guess where the system has no region to give. The Dart
  /// side reads that as "fall back to the language", which is what it did
  /// before it could ask at all.
  private static func current() -> String? {
    if #available(macOS 13.0, iOS 16.0, *) {
      return Locale.current.region?.identifier
    }
    return (Locale.current as NSLocale).object(forKey: .countryCode) as? String
  }
}
